store/src/java/com/zimbra/qa/unittest/TestUtil.java

```diff
@@ -555,9 +617,52 @@ private static ZMimePart getPart(ZMimePart mimeStructure, String name) {
                 ZimbraLog.test.debug("sleep got interrupted", e);
             }
         }
+        ZimbraLog.test.debug("helpWait4Msgs finished for mbox='%s' query='%s' %s", mbox.getName(), query,
+                ZimbraLog.elapsedTime(start, System.currentTimeMillis()));
+        return msgs;
+    }
+
+    /**
+     * @param numMsgsExpected
+     *            - 0 means don't expect a message to arrive before timeout_millis
+     * @param timeout_millis
+     * @throws ServiceException
+     */
+    public static List<ZMessage> waitForMessages(ZMailbox mbox, String query, int numMsgsExpected, int timeout_millis)
+            throws ServiceException {
+        long start = System.currentTimeMillis();
+        long finishBy = start + timeout_millis;
+        List<ZMessage> msgs;
+        // Just wait for a third of the time initially.  If hasn't arrived in that time, poke postfix to process
+        // its queue and wait up to the rest of the time
+        msgs = helpWait4Msgs(mbox, query, numMsgsExpected, timeout_millis / 3);
+        if ((numMsgsExpected > 0) && (msgs.size() == numMsgsExpected)) {
+            return msgs;
+        }
+        if (numMsgsExpected > 0) {
+            /* One last try.  See if got stuck in Postfix mailq - seen mailq entries like this before now:
+             * E980B183AF8     2622 Thu Mar 30 14:49:54  user1@example.com
+             * (delivery temporarily suspended: connect to example.com[XXX.XXX.XXX.XXX]:7025: Connection refused)
+             *                           user1@example.com
+             * and:
+             * DD410183B37     2619 Thu Mar 30 15:59:31  user1@example.com
+             * (host example.com[XXX.XXX.XXX.XXX] said:
+             *     451 4.0.0 Temporary message delivery failure try again (in reply to end of DATA command))
+             *                           user1@example.com
+             */
+            if (postfixFlushDeferredMailQueue(AccountTestUtil.getServer(mbox.getName()),
+                    finishBy - System.currentTimeMillis())) {
```

Comment:
Been round in circles on this.  I've come to the conclusion that this is probably as good as it gets with the current code.  If there is an issue, it is likely to be that the remote queue handling code works on **Server** objects and assumes that Provisioning.A_zimbraServiceHostname attribute for that Server is the right host to execute postfix queue handling commands on.